lib/AST/ASTPrinter.cpp

```diff
@@ -3374,6 +3418,10 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
       : Printer(Printer), Options(PO) {}
 
   void visit(Type T) {
+    if (Options.SelfSubstitutionType && T->is<DynamicSelfType>()) {
```

Comment:
Hmm... this is a pretty narrow solution. It won't work for, e.g.,:

    static func +(lhs: Self, rhs: Self) -> Self

right?